replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/WorkerThreadPool.java

```diff
@@ -0,0 +1,160 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.replication.api.impl.worker;
+
+import com.connexta.replication.api.NodeAdapter;
+import com.connexta.replication.api.data.Site;
+import com.connexta.replication.api.impl.NodeAdapters;
+import com.connexta.replication.api.queue.Queue;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Manages a site's set of workers which in turn run workers.
+ */
+public class WorkerThreadPool {
+
+  private final Object lock = new Object();
+
+  private boolean shutdown = false;
+
+  private final List<Worker> workers = Collections.synchronizedList(new ArrayList<>());
+
+  private final NodeAdapters nodeAdapters;
+
+  private final Site localSite;
+
+  private final Site destinationSite;
+
+  private final Queue queue;
+
+  private AtomicInteger size = new AtomicInteger(0);
+
+  /**
+   * Creates a new worker thread pool for the given site and spins up a number of workers equal to
+   * the size passed in.
+   *
+   * @param queue queue which workers will poll tasks from
+   * @param size the initialize size of the pool
+   * @param localSite the local site
+   * @param remoteSite the remote site corresponding to the queue
+   * @param nodeAdapters factory for creating node adapters
+   */
+  public WorkerThreadPool(Queue queue, int size,
+      Site localSite,
+      Site remoteSite,
+      NodeAdapters nodeAdapters) {
+    this.queue = queue;
+    this.localSite = localSite;
+    this.destinationSite = remoteSite;
+    this.nodeAdapters = nodeAdapters;
+    setSize(size);
+  }
+
+  /**
+   * Sets the size of this thread pool. If the current size exceeds the new size, then workers
+   * workers will be interrupted until the new size is achieved. Workers that are currently not
+   * processing will be prioritized for cleanup before workers that are processing. If all workers
+   * are processing or there is a remaining amount that needs to be interrupted, then they will be
+   * stopped in the order they are encountered until the amount of workers un-interrupted is equal
+   * to the new size.
+   *
+   * If the new size is greater than the current size, then newSize-currentSize workers will be spun
+   * up.
+   *
+   * @param size the size of this thread pool
+   * @throws IllegalStateException if this thread pool has been shutdown
+   */
+  public void setSize(int size) {
+    synchronized (lock) {
+      if (shutdown) {
+        throw new IllegalStateException("This thread is shutdown");
+      }
+    }
+
+    final int currentSize = this.getSize();
```

Comment:
❗️ The whole operations of checking the size and updating it should be done under lock to prevent other threads from possibly calling this method while you are affecting the state of the object stored in multiple attributes.